File: src/data/BytesData.cpp
//
// Created by Andrei Costinescu (andreicostinescu96@gmail.com) on 25-Apr-21.
//

#include <comm/data/BytesData.h>
#include <comm/socket/utils.h>
#include <cassert>
#include <iostream>

using namespace comm;
using namespace std;

const int BytesData::headerSize = sizeof(int);

BytesData::BytesData(int size) : data(size), expectedDataSize(0) {}

BytesData::~BytesData() {
    this->data.reset();
}

MessageType BytesData::getMessageType() {
    return MessageType::BYTES;
}

bool BytesData::serialize(Buffer *buffer, int start, bool forceCopy, bool verbose) {
    switch (this->serializeState) {
        case 0: {
            // header
            buffer->setBufferContentSize(start + BytesData::headerSize);
            buffer->setInt((int) this->data.getBufferContentSize(), start);
            if (verbose) {
                const char *dataBuffer = buffer->getBuffer();
                cout << "buffer int content: " << (int) dataBuffer[start] << " " << (int) dataBuffer[start + 1] << " "
                     << (int) dataBuffer[start + 2] << " " << (int) dataBuffer[start + 3] << endl;
            }
            this->serializeState = 1;
            return false;
        }
        case 1: {
            if (forceCopy) {
                buffer->setData(this->getBuffer(), this->getBufferSize());
            } else {
                if (start != 0) {
                    throw runtime_error("Can not set a reference to data not starting at the first position!");
                }
                buffer->setConstReferenceToData(this->getBuffer(), this->getBufferSize());
            }
            this->serializeState = 0;
            return true;
        }
        default: {
            (*cerror) << "Impossible serialize state... " << this->serializeState << endl;
            this->serializeState = 0;
            return false;
        }
    }
}

int BytesData::getExpectedDataSize() const {
    switch (this->deserializeState) {
        case 0: {
            return BytesData::headerSize;
        }
        case 1: {
            return this->expectedDataSize;
        }
        default : {
            throw runtime_error("Impossible deserialize state... " + to_string(this->deserializeState));
        }
    }
}

bool BytesData::deserialize(Buffer *buffer, int start, bool, bool verbose) {
    switch (this->deserializeState) {
        case 0: {
            this->expectedDataSize = buffer->getInt(start);
            this->deserializeState = 1;
            return false;
        }
        case 1: {
            assert ((buffer->getBufferContentSize() - start) == this->expectedDataSize);
            this->data.setData(buffer->getBuffer() + start, this->expectedDataSize);
            this->deserializeState = 0;
            return true;
        }
        default: {
            (*cerror) << "Impossible deserialize state... " << this->deserializeState << endl;
            this->resetDeserializeState();
            return false;
        }
    }
}

void BytesData::reset() {
    this->data.reset();
    this->expectedDataSize = 0;
}

void BytesData::setData(char *_data, unsigned int dataSize, int start) {
    this->data.setData((const char *) _data, dataSize, start);
}

void BytesData::setData(const char *_data, unsigned int dataSize, int start) {
    this->data.setData(_data, dataSize, start);
}

void BytesData::setReferenceToData(char *_data, unsigned int dataSize) {
    this->data.setReferenceToData(_data, dataSize);
}

void BytesData::setChar(char _data, int position) {
    this->data.setChar(_data, position);
}

void BytesData::setShort(short _data, int position) {
    this->data.setShort(_data, position);
}

void BytesData::setInt(int _data, int position) {
    this->data.setInt(_data, position);
}

void BytesData::setLongLong(long long _data, int position) {
    this->data.setLongLong(_data, position);
}

void BytesData::setFloat(float _data, int position) {
    this->data.setFloat(_data, position);
}

void BytesData::setDouble(double _data, int position) {
    this->data.setDouble(_data, position);
}

char BytesData::getChar(int position) {
    return this->data.getChar(position);
}

short BytesData::getShort(int position) {
    return this->data.getShort(position);
}

int BytesData::getInt(int position) {
    return this->data.getInt(position);
}

long long BytesData::getLongLong(int position) {
    return this->data.getLongLong(position);
}

float BytesData::getFloat(int position) {
    return this->data.getFloat(position);
}

double BytesData::getDouble(int position) {
    return this->data.getDouble(position);
}

bool BytesData::empty() const {
    return this->data.empty();
}

const char *BytesData::getBuffer() {
    return this->data.getBuffer();
}

uint64_t BytesData::getBufferSize() const {
    uint64_t size = this->data.getBufferContentSize();
    assert (this->deserializeState != 1 || size == this->expectedDataSize);
    return size;
}

File: csharp/data/BytesData.cs
using System;
using System.Diagnostics;

namespace DesignAIRobotics.Comm.data {
    class BytesData : CommunicationData {
        const int headerSize = sizeof(int);

        public BytesData(int size) {
            this.data = new utils.Buffer((ulong) size);
            this.expectedDataSize = 0;

        }

        ~BytesData() {
            this.data.reset();
        }

        public override MessageType getMessageType() {
            return MessageType.BYTES;
        }

        public override bool serialize(utils.Buffer buffer, int start, bool forceCopy, bool verbose) {
            switch (this.serializeState) {
                case 0: {
                    // header
                    buffer.setBufferContentSize((ulong) start + headerSize);
                    buffer.setInt((int) this.data.getBufferContentSize(), start);
                    if (verbose) {
                        byte[] dataBuffer = buffer.getBuffer();
                        Console.WriteLine("buffer int content: " + (int) dataBuffer[0] + " " + (int) dataBuffer[1] + " " + (int) dataBuffer[2] + " " + (int) dataBuffer[3]);
                    }
                    this.serializeState = 1;
                    return false;
                }
                case 1: {
                    if (forceCopy) {
                        buffer.setData(this.getBuffer(), this.getBufferSize(), (ulong) start);
                    } else {
                        if (start != 0) {
                            throw new System.Exception("Can not set a reference to data not starting at the first position!");
                        }
                        buffer.setReferenceToData(this.getBuffer(), this.getBufferSize());
                    }
                    this.serializeState = 0;
                    return true;
                }
                default: {
                    Console.WriteLine("Impossible serialize state... " + this.serializeState);
                    this.serializeState = 0;
                    return false;
                }
            }
        }

        public override ulong getExpectedDataSize() {
            switch (this.deserializeState) {
                case 0: {
                    return headerSize;
                }
                case 1: {
                    return (ulong) this.expectedDataSize;
                }
                default: {
                    throw new System.Exception("Impossible deserialize state... " + this.deserializeState);
                }
            }
        }

        public override bool deserialize(utils.Buffer buffer, int start, bool forceCopy, bool verbose) {
            switch (this.deserializeState) {
                case 0: {
                    this.expectedDataSize = buffer.getInt(start);
                    this.deserializeState = 1;
                    return false;
                }
                case 1: {
                    Debug.Assert((buffer.getBufferContentSize() - (ulong) start) == (ulong) this.expectedDataSize);
                    this.data.setData(buffer.getBuffer(), (ulong) this.expectedDataSize, (ulong) start);
                    this.deserializeState = 0;
                    return true;
                }
                default: {
                    Console.WriteLine("Impossible deserialize state... " + this.deserializeState);
                    this.resetDeserializeState();
                    return false;
                }
            }
        }

        public void reset() {
            this.data.reset();
            this.expectedDataSize = 0;
        }

        public void setData(byte[] _data, uint dataSize, ulong start) {
            this.data.setData(_data, dataSize, start);
        }

        public void setReferenceToData(byte[] _data, uint dataSize) {
            this.data.setReferenceToData(_data, dataSize);
        }

        public void setByte(byte _data, int position) {
            this.data.setByte(_data, position);
        }

        public void setShort(short _data, int position) {
            this.data.setShort(_data, position);
        }

        public void setInt(int _data, int position) {
            this.data.setInt(_data, position);
        }

        public void setLongLong(long _data, int position) {
            this.data.setLongLong(_data, position);
        }

        public void setFloat(float _data, int position) {
            this.data.setFloat(_data, position);
        }

        public void setDouble(double _data, int position) {
            this.data.setDouble(_data, position);
        }

        public byte getByte(int position) {
            return this.data.getByte(position);
        }

        public short getShort(int position) {
            return this.data.getShort(position);
        }

        public int getInt(int position) {
            return this.data.getInt(position);
        }

        public long getLongLong(int position) {
            return this.data.getLongLong(position);
        }

        public float getFloat(int position) {
            return this.data.getFloat(position);
        }

        public double getDouble(int position) {
            return this.data.getDouble(position);
        }

        public bool empty() {
            return this.data.empty();
        }

        public byte[] getBuffer() {
            return this.data.getBuffer();
        }

        public ulong getBufferSize() {
            ulong size = this.data.getBufferContentSize();
            Debug.Assert(this.deserializeState != 1 || size == (ulong) this.expectedDataSize);
            return size;
        }

        private utils.Buffer data;
        private int expectedDataSize;
    }
}
